### ISSUE TITLE
Motivation for deviating from Ecto 

### ISSUE THREAD

**Issue by venkatd** [arpieb/spiro#1]
Hi I just spoke with @florinpatrascu about writing a Neo4j Ecto adapter and he mentioned that you opted to create a completely new project.

Could you talk about some of the issues you encountered with going down the Ecto route? Also, curious to hear about your plans for the Graph DSL :) 

Personally I feel that besides basic CRUD (including relationships) and lookup statements, it is best to keep the DSL as close to the actual language as possible. For example when working with neo4j, I'd prefer to see the actual cypher query rather than something I'd have to convert to neo4j in my head.

That being said, it would be great to put together a neatly composable DSL. Would love to hear your thoughts!

**Comment by arpieb** [comment]
Right, so as @florinpatrascu had mentioned in his post, when he tried to jam edge traversals and some more advanced graph-specific matching login into Ecto, he started running into DSL limitations and assumptions regarding how Ecto's SQL-specific query system handles schemas and query construction.  Instead of effectively "crippling" graph database queries, my thought is to build an Ecto analogue specifically for graph databases so we don't lose anything in building adapters for it.

Roadmap is in the 0.x branch's README file, but in summary I'm going to implement the TinkerPop3 Gremlin Structure's DSL in Elixir in such a way that it can be used to build database-specific adapters (a la Ecto) and even talk directly to Gremlin-enabled datastores.

The reasoning is as follows:
* TinkerPop's DSL is based on a transformative query pipeline, which totally jives with Elixir pipes.  And it's not a huge leap for Ecto users.
* TP's already proven that it can support a wide range of not only graph databases, but languages (Clojure and Scala are represented, so we know it can play nicely with functional languages).
* It's mature - version 3, and an Apache Incubator project.  This isn't some cooked-up-over-lattes DSL, it's been tested and integrates with graph compute systems as well.
* Florin has done a bang-up job building a very specific package for Neo4j, but I'd like to have something like Ecto that can be extended by the community.

It's my first Elixir project, so a bit of an undertaking what with work and grad school, but I'd appreciate any help you'd like to contribute!  Currently trying to map Ecto general-purpose architecture in (e.g. pool management, adapter config/selection, field data types) and modules that map to Gremlin structures.  First adapter target is digraph since it's baked into Erlang+Elixir *and* it gives us a nice datastore for unit tests of general logic.

-R

**Comment by arpieb** [comment]
Oh, and it's not a "deviation" from Ecto - I have **absolutely no plans** to replicate any Ecto non-graph databases in Spiro.  This project is strictly for building a DSL-based integration framework for graph databases.  It is meant to complement Ecto, not be a competitor.

**Comment by venkatd** [comment]
Hi @arpieb sorry for the delayed response it has been a busy week. I think TinkerPop looks like a great approach to this given it's wide support for graph databases.

For (selfish) reasons, I'd like to focus on building something optimized uses neo4j. Here the considerations:
- We are taking advantage of neo4j specific capabilties and don't have any plans to switch databases
- I find the neo4j Cypher query syntax more natural -- it also has some similarities that lend itself to a functional style. The Cypher syntax is largely based on pattern matching and allows you to pipe queries into each other using the `WITH` syntax
- There are some neo4j specific queries that I'm leaning on

That being said, would love to give feedback along the way and once you have a reference implementation, I wouldn't mind helping building a neo4j integration.

**Comment by arpieb** [comment]
@venkatd in that case i'd definitely recommend working with Florin on the neo4j_sips project as it's laser-focused on supporting Neo4j features, whereas the goal of Spiro is to come up with something more generic that could support applications that allow for datastore backends to be swapped out.

I would appreciate any and all help you can provide when I get to the Neo4j interface, so will keep you posted!  Thanks for the interest!

**Comment by venkatd** [comment]
Sounds good @arpieb!
